networking/p2p/src/main/java/tech/pegasys/teku/networking/p2p/connection/ConnectionManager.java

```diff
@@ -91,17 +94,36 @@ protected SafeFuture<?> doStart() {
     synchronized (this) {
       staticPeers.forEach(this::createPersistentConnection);
     }
-    periodicPeerSearch =
-        asyncRunner.runWithFixedDelay(
-            this::searchForPeers,
-            DISCOVERY_INTERVAL,
-            error -> LOG.error("Error while searching for peers", error));
-    connectToBestPeers(emptyList());
-    searchForPeers();
+    createRecurrentSearchTask();
     peerConnectedSubscriptionId = network.subscribeConnect(this::onPeerConnected);
     return SafeFuture.COMPLETE;
   }
 
+  private void createRecurrentSearchTask() {
+    searchForPeers().alwaysRun(this::createNextSearchPeerTask).finish(this::logSearchError);
+  }
+
+  private void logSearchError(final Throwable throwable) {
+    LOG.error("Error while searching for peers", throwable);
+  }
+
+  private void createNextSearchPeerTask() {
+    if (warmupDiscovery.get()) {
+      // Retry fast until first peer connection is initiated, which means Discovery is warmed up
+      this.periodicPeerSearch =
```

Comment:
The condition we want is that we retry frequently until the first peer is found, so I'd just make the if condition `if (network.getPeerCount() > 0)`. That would then solve the issue you mention of a 2 node network as well.

It's possible we'll get an inbound peer quickly and so switch to the slower searching before discovery is fully ready but if we already have a peer that's not a big deal as we will already be able to start catching up again.